src/Validation/Validator.php

```diff
@@ -2184,6 +2184,11 @@ public function integer(string $field, ?string $message = null, Closure|string|n
      */
     public function array(string $field, ?string $message = null, Closure|string|null $when = null)
     {
+        $message ??= 'The provided value must be an array';
+        if ($this->_useI18n) {
+            $message = __d('cake', 'The provided value must be an array');
```

Comment:
You are potentially overriding user provided value here.

Reply:
Yeah we can't overwrite messages like this as we'll replace developer provided messages here.

Reply:
That wasn't intended. Copy & Pasta.

Which style for checking the user provided message do you prefer?

1. With encompassing  null check:

``` php
        if ($message === null) {
            $message = 'The provided value must be scalar';
            if ($this->_useI18n) {
                $message = __d('cake', 'The provided value must be scalar');
            }
        }
```

2. With null coalescing assignment operators:

``` php
        $message ??= 'The provided value must be scalar';
        if ($this->_useI18n) {
            $message ??= __d('cake', 'The provided value must be scalar');
        }
```

I think the first is more traditional and uses less operations.

3. With encompassing  null check & if/else trending to the use of I18n:

``` php
        if ($message === null) {
            if ($this->_useI18n) {
                $message = __d('cake', 'The provided value must be scalar');
            } else {
                $message = 'The provided value must be scalar';
            }
        }
```

4. With encompassing  null check & if/else trending to the non-use of I18n:

``` php
        if ($message === null) {
            if (!$this->_useI18n) {
                $message = 'The provided value must be scalar';
            } else {
                $message = __d('cake', 'The provided value must be scalar');
            }
        }
```

From an optical view-point (pun wasn't intended) , I would go with the fourth.  For readability the first one, even though it might be a little less efficient (for I18n-enabled applications).

Reply:
```php
         $message ??= $this->_useI18n
                 ? __d('cake', 'The provided value must be scalar')
                 :  'The provided value must be scalar';
```

Reply:
If you end up going with a more verbose implementation could we extract it into a protected method so we don't end up with multiple copies. You might need to use validation message templates to make this work. Perhaps a message like `The provided value must be {0}`

Reply:
> If you end up going with a more verbose implementation could we extract it into a protected method so we don't end up with multiple copies. 

Certainly. Extracting it into a protected method was actually my idea, too.

> You might need to use validation message templates to make this work. Perhaps a message like `The provided value must be {0}` would work?

I'm not sure if using such a split sentence is good from a I18n point of view. In some languages one would probably word the whole sentense differently in some cases instead of just replacing one term with another.
That is why I would go with full sentences.

Reply:
Went with the last style.